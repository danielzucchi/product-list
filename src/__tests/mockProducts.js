const mockProducts = [
  {
    id: '1',
    date: '2020-11-11T03:23:31.984Z',
    name: 'Rowena Jerde',
    img: 'https://s3.amazonaws.com/uifaces/faces/twitter/a_harris88/128.jpg',
    sold: false,
    price: '51.00',
    brand: 'Fantastic Concrete Computer',
    description:
      "You can't copy the capacitor without generating the digital SAS hard drive!",
    seller: 'Frederic',
  },
  {
    id: '2',
    date: '2020-11-11T03:25:47.608Z',
    name: 'Walter Nicolas',
    img: 'https://s3.amazonaws.com/uifaces/faces/twitter/uxalex/128.jpg',
    sold: true,
    price: '348.00',
    brand: 'Handcrafted Metal Car',
    description:
      "I'll override the open-source HTTP protocol, that should interface the COM program!",
    seller: 'Lonnie',
  },
  {
    id: '3',
    date: '2020-11-11T08:07:26.985Z',
    name: 'Justina Heaney',
    img: 'https://s3.amazonaws.com/uifaces/faces/twitter/juangomezw/128.jpg',
    sold: false,
    price: '461.00',
    brand: 'Unbranded Frozen Pizza',
    description:
      'Try to synthesize the PNG monitor, maybe it will parse the primary pixel!',
    seller: 'Calista',
  },
  {
    id: '4',
    date: '2020-11-10T17:50:08.753Z',
    name: 'Mrs. Elizabeth Tremblay',
    img: 'https://s3.amazonaws.com/uifaces/faces/twitter/operatino/128.jpg',
    sold: false,
    price: '71.00',
    brand: 'Handcrafted Rubber Chair',
    description:
      "overriding the transmitter won't do anything, we need to back up the primary HTTP driver!",
    seller: 'Kirk',
  },
]

export default mockProducts
