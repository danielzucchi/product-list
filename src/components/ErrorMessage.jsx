import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import exclamationIcon from '../assets/exclamation.svg'

const ErrorWrapper = styled.div`
  display: flex;
  align-items: center;
  width: 60%;
  background-color: #fdecea;
  min-height: 32px;
  border: none;
  margin: 10px auto;
  padding: 6px 10px;
  border-radius: 5px;
`

const Image = styled.img`
  float: left;
  width: 16px;
  height: 16px;
  margin-right: 10px;
`

const ContentText = styled.span`
  font-size: 13px;
`

const ErrorMessage = ({ error }) => (
  <ErrorWrapper>
    <Image src={exclamationIcon} alt="error" />
    <ContentText>{error}</ContentText>
  </ErrorWrapper>
)

ErrorMessage.propTypes = {
  error: PropTypes.string.isRequired,
}

export default ErrorMessage
