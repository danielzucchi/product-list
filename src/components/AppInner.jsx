import React, { useState } from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import LoadingSpinner from './LoadingSpinner/LoadingSpinner'
import ErrorMessage from './ErrorMessage'
import AppBar from './AppBar'
import ProductTileList from './ProductTileList'

const ProductListContainer = styled.div`
  margin: 50px auto;
  width: 70%;
`

const LoadingWrapper = styled.div`
  text-align: center;
  margin: 10% auto;
`

const SubMenuBar = styled.div`
  width: 100%;
  margin: 20px auto;
  height: 40px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
`

const ResultsTitle = styled.div`
  font-size: 18px;
  font-weight: bold;
`

const HideButton = styled.button`
  background-color: ${({ hideSold }) => (hideSold ? '#3e3e3f' : '#fff')};
  padding: 10px 20px;
  text-align: center;
  text-transform: uppercase;
  background-size: 200% auto;
  color: ${({ hideSold }) => (hideSold ? '#fff' : '##3e3e3f')};
  display: block;
  outline: none;
  border: 1px solid #3e3e3f;
  border-radius: 5px;
`

const AppInner = ({ isLoading, products, error }) => {
  const [hideSold, setHideSold] = useState(false)
  const [likedProducts, setLiked] = useState([])

  const likedProductsList = products
    ? products
        .filter((product) =>
          likedProducts.find((productId) => productId === product.id),
        )
        .map((product) => product.brand)
    : []

  const toggleLiked = (id) => {
    if (likedProducts.includes(id)) {
      setLiked(likedProducts.filter((productId) => id !== productId))
    } else {
      setLiked(likedProducts.concat(id))
    }
  }

  return (
    <>
      <AppBar likedProducts={likedProductsList} />
      {error && <ErrorMessage error={error} />}
      <ProductListContainer>
        <SubMenuBar>
          <ResultsTitle>Results</ResultsTitle>
          <HideButton
            type="button"
            hideSold={hideSold}
            onClick={() => setHideSold(!hideSold)}
          >
            {hideSold ? 'Show sold items' : 'Hide sold items'}
          </HideButton>
        </SubMenuBar>
        {isLoading && (
          <LoadingWrapper>
            <LoadingSpinner />
          </LoadingWrapper>
        )}
        {products && (
          <ProductTileList
            likedProducts={likedProducts}
            products={products}
            hideSold={hideSold}
            toggleLiked={toggleLiked}
          />
        )}
      </ProductListContainer>
    </>
  )
}

AppInner.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  products: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      date: PropTypes.string,
      name: PropTypes.string,
      img: PropTypes.string,
      sold: PropTypes.bool,
      price: PropTypes.string,
      brand: PropTypes.string,
      description: PropTypes.string,
      seller: PropTypes.string,
    }),
  ),
  error: PropTypes.string,
}

AppInner.defaultProps = {
  products: null,
  error: null,
}

export default AppInner
