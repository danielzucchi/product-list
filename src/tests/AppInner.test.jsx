import React from 'react'
import { shallow } from 'enzyme'
import AppInner from '../components/AppInner'
import LoadingSpinner from '../components/LoadingSpinner/LoadingSpinner'
import ErrorMessage from '../components/ErrorMessage'
import mockProducts from './mockProducts'

describe('AppInner tests', () => {
  it('when all props are presentmatches snapshot', () => {
    const wrapper = shallow(
      <AppInner isLoading={false} products={mockProducts} error={null} />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('shows loading spinner when loading', () => {
    const wrapper = shallow(<AppInner isLoading />)
    const loadingSpinner = wrapper.find(LoadingSpinner)
    expect(loadingSpinner).toBeDefined()
  })

  it('shows error when API throws error', () => {
    const wrapper = shallow(
      <AppInner
        isLoading={false}
        error="An error has occurred"
        products={null}
      />,
    )
    const error = wrapper.find(ErrorMessage)
    expect(error).toHaveLength(1)
  })
})
