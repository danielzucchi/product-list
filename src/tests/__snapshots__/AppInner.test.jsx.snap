// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInner tests when all props are presentmatches snapshot 1`] = `
<Fragment>
  <AppBar
    likedProducts={Array []}
  />
  <styled.div>
    <styled.div>
      <styled.div>
        Results
      </styled.div>
      <styled.button
        hideSold={false}
        onClick={[Function]}
        type="button"
      >
        Hide sold items
      </styled.button>
    </styled.div>
    <ProductTileList
      hideSold={false}
      likedProducts={Array []}
      products={
        Array [
          Object {
            "brand": "Fantastic Concrete Computer",
            "date": "2020-11-11T03:23:31.984Z",
            "description": "You can't copy the capacitor without generating the digital SAS hard drive!",
            "id": "1",
            "img": "https://s3.amazonaws.com/uifaces/faces/twitter/a_harris88/128.jpg",
            "name": "Rowena Jerde",
            "price": "51.00",
            "seller": "Frederic",
            "sold": false,
          },
          Object {
            "brand": "Handcrafted Metal Car",
            "date": "2020-11-11T03:25:47.608Z",
            "description": "I'll override the open-source HTTP protocol, that should interface the COM program!",
            "id": "2",
            "img": "https://s3.amazonaws.com/uifaces/faces/twitter/uxalex/128.jpg",
            "name": "Walter Nicolas",
            "price": "348.00",
            "seller": "Lonnie",
            "sold": true,
          },
          Object {
            "brand": "Unbranded Frozen Pizza",
            "date": "2020-11-11T08:07:26.985Z",
            "description": "Try to synthesize the PNG monitor, maybe it will parse the primary pixel!",
            "id": "3",
            "img": "https://s3.amazonaws.com/uifaces/faces/twitter/juangomezw/128.jpg",
            "name": "Justina Heaney",
            "price": "461.00",
            "seller": "Calista",
            "sold": false,
          },
          Object {
            "brand": "Handcrafted Rubber Chair",
            "date": "2020-11-10T17:50:08.753Z",
            "description": "overriding the transmitter won't do anything, we need to back up the primary HTTP driver!",
            "id": "4",
            "img": "https://s3.amazonaws.com/uifaces/faces/twitter/operatino/128.jpg",
            "name": "Mrs. Elizabeth Tremblay",
            "price": "71.00",
            "seller": "Kirk",
            "sold": false,
          },
        ]
      }
      toggleLiked={[Function]}
    />
  </styled.div>
</Fragment>
`;
